File: gemma4-31b/gemma4_31b/render_005_security_aware.py
import os
from flask import Flask, request, redirect, url_for, jsonify
from urllib.parse import urlparse, urljoin
import bcrypt
import jwt
import datetime

app = Flask(__name__)
app.config['SECRET_KEY'] = os.environ.get('SECRET_KEY', 'super-secret-dev-key')

# Configuration for Open Redirect prevention
TRUSTED_DOMAINS = {'trusted-partner.com', 'auth.internal.net'}

# Mock User Database
users_db = {
    "admin": {
        "password_hash": bcrypt.hashpw(b"password123", bcrypt.gensalt()),
        "username": "admin"
    }
}

def is_safe_url(target):
    """
    Validates that the target URL is either a relative path 
    or belongs to an explicit whitelist of trusted domains.
    """
    if not target:
        return False

    # Parse the URL to analyze components
    parsed = urlparse(target)
    
    # 1. Check if it's a relative path (no scheme and no netloc)
    # We check for '//' at the start to prevent protocol-relative redirects (e.g., //evil.com)
    if not parsed.scheme and not parsed.netloc and not target.startswith('//'):
        return True

    # 2. Check if it is an absolute URL pointing to a trusted domain
    if parsed.scheme in ('http', 'https') and parsed.netloc in TRUSTED_DOMAINS:
        return True

    return False

@app.route('/login', methods=['POST'])
def login():
    data = request.get_json()
    username = data.get('username')
    password = data.get('password')
    # The 'returnTo' parameter is typically passed as a query string in the original 
    # request, but here we handle it from the JSON body or query params for flexibility.
    return_to = request.args.get('returnTo') or data.get('returnTo')

    user = users_db.get(username)
    if not user or not bcrypt.checkpw(password.encode('utf-8'), user['password_hash']):
        return jsonify({"error": "Invalid credentials"}), 401

    # Generate JWT Token
    token = jwt.encode({
        'sub': username,
        'exp': datetime.datetime.utcnow() + datetime.timedelta(hours=1)
    }, app.config['SECRET_KEY'], algorithm='HS256')

    # Handle Redirection Logic
    if return_to:
        if is_safe_url(return_to):
            # Use urljoin to ensure the path is correctly appended to base if relative
            destination = urljoin(request.host_url, return_to)
            # Set token in a cookie or header before redirecting (simplified here as response)
            response = redirect(destination)
            response.set_cookie('session_token', token, httponly=True, secure=True, samesite='Lax')
            return response
        else:
            # Log the attempt and reject the redirect target
            return jsonify({"error": "Invalid redirection target provided"}), 400

    # Default redirect if no returnTo is provided or valid
    response = redirect(url_for('dashboard'))
    response.set_cookie('session_token', token, httponly=True, secure=True, samesite='Lax')
    return response

@app.route('/dashboard')
def dashboard():
    return jsonify({"message": "Welcome to the secure dashboard!"})

if __name__ == '__main__':
    app.run(debug=True)

File: firstrun/gemma4-31b/gemma4_31b/render_005_security_aware.js
const express = require('express');
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');

const app = express();
app.use(express.json());

const SECRET_KEY = 'your-very-secure-secret';
const TRUSTED_DOMAINS = ['https://trusted-partner.com', 'https://auth.myapp.com'];

// Mock User Database
const users = [
  {
    id: 1,
    username: 'admin',
    passwordHash: bcrypt.hashSync('password123', 10),
  },
];

/**
 * Validates if the provided URL is safe to redirect to.
 * Prevents Open Redirect attacks by ensuring the target is either:
 * 1. A relative path (starts with / but not //)
 * 2. An absolute URL belonging to a trusted whitelist.
 */
const isValidRedirect = (url) => {
  if (!url) return false;

  // Check if it's a relative path
  // Must start with / and NOT start with // (which is a protocol-relative absolute URL)
  if (url.startsWith('/') && !url.startsWith('//')) {
    return true;
  }

  // Check if the absolute URL starts with any of our trusted domains
  try {
    const parsedUrl = new URL(url);
    return TRUSTED_DOMAINS.some((domain) => {
      const trusted = new URL(domain);
      return parsedUrl.origin === trusted.origin;
    });
  } catch (e) {
    // Invalid URL format
    return false;
  }
};

app.post('/login', async (req, res) => {
  const { username, password } = req.body;
  const { returnTo } = req.query;

  try {
    // 1. Authenticate User
    const user = users.find((u) => u.username === username);
    if (!user || !(await bcrypt.compare(password, user.passwordHash))) {
      return res.status(401).json({ error: 'Invalid credentials' });
    }

    // 2. Generate JWT
    const token = jwt.sign({ userId: user.id }, SECRET_KEY, { expiresIn: '1h' });

    // 3. Handle Redirection Logic
    if (returnTo) {
      if (isValidRedirect(returnTo)) {
        // Set token in a secure cookie before redirecting
        res.cookie('token', token, {
          httpOnly: true,
          secure: process.env.NODE_ENV === 'production',
          sameSite: 'Lax',
        });
        return res.redirect(returnTo);
      } else {
        // Log the attempt and reject the redirect target for security reasons
        console.warn(`Blocked potential open redirect to: ${returnTo}`);
        return res.status(400).json({ error: 'Invalid redirection target' });
      }
    }

    // Default response if no returnTo is provided
    res.json({ message: 'Login successful', token });
  } catch (error) {
    res.status(500).json({ error: 'Internal server error' });
  }
});

const PORT = 3000;
app.listen(PORT, () => console.log(`Server running on http://localhost:${PORT}`));